lib/restforce/db/timestamp_cache.rb

```diff
@@ -0,0 +1,76 @@
+module Restforce
+
+  module DB
+
+    # Restforce::DB::TimestampCache serves to cache the timestamps of the most
+    # recent known updates to records through the Restforce::DB system. It
+    # allows for more intelligent decision-making regarding what constitutes
+    # "stale" data during a synchronization.
+    class TimestampCache
+
+      # Public: Initialize a new Restforce::DB::TimestampCache.
+      def initialize
+        reset
+      end
+
+      # Public: Add a known update timestamp to the cache for the passed object.
+      #
+      # instance - A Restforce::DB::Instances::Base.
+      #
+      # Returns an Array of Restforce::DB::Instances::Base.
+      def update(instance)
+        @cache[key_for(instance)] = instance.last_update
+      end
+
+      # Public: Get the most recently-stored timestamp for the passed object.
+      #
+      # instance - A Restforce::DB::Instances::Base.
+      #
+      # Returns a Time or nil.
+      def timestamp(instance)
+        key = key_for(instance)
+        @cache.fetch(key) { @last_cache[key] }
+      end
+
+      # Public: Has the passed instance been modified since the last known
+      # system-triggered update? This accounts for changes possibly introduced
+      # by callbacks and triggers.
+      #
+      # instance - A Restforce::DB::Instances::Base.
+      #
+      # Returns a Boolean.
+      def changed?(instance)
+        return true unless instance.updated_internally?
+
+        last_update = timestamp(instance)
+        return true unless last_update
+
+        instance.last_update > last_update
+      end
+
+      # Public: Reset the cache. Expires the previously-cached timestamps, and
+      # retires the currently-cached timestamps to ensure that they are only
+      # factored into the current synchronization run.
+      #
+      # Returns nothing.
+      def reset
+        @last_cache = @cache || {}
```

Comment:
`@last_cache` is actually the entire reason this works; after one synchronization loop, we need to know for the following loop whether or not the record has changed since our intentional modification. Without falling back to the "retired" values (see `#timestamp`'s implementation), subsequents loops would have no context into what happened in the previous run.